File: from-0101-to-0200/0143-reorder-list/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func reorderList(head *ListNode)  {
    idx := 0
    p := head
    nodes := make(map[int]*ListNode)
    for p != nil {
        nodes[idx] = p
        p = p.Next
        idx += 1
    }

    left, right := 1, idx - 1
    p = head
    for left < right {
        p.Next = nodes[right]
        p = p.Next
        p.Next = nodes[left]
        p = p.Next
        right -= 1
        left += 1
    }

    if left == right {
        p.Next = nodes[left]
        p = p.Next
    }
    p.Next = nil
}

File: from-0101-to-0200/0143-reorder-list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        nodes = []
        p = head.next
        while p:
            nodes.append(p)
            p = p.next
            
        left, right = 0, len(nodes)-1
        p = head
        
        while left <= right:
            p.next = nodes[right]
            p = p.next
            right -= 1
            p.next = nodes[left]
            p = p.next
            left += 1
            
        p.next = None
        